### PR TITLE
State refactor

### PR DESCRIPTION
**Pull request by Ninjatosba**

### CODE REVIEW

File: contracts/stream/src/contract.rs [hunk 1]
```diff
@@ -1,10 +1,13 @@
-use crate::helpers::{check_name_and_url, get_decimals};
+use core::str;
+use std::env;
```
Comment (orkunkl) [inline contracts/stream/src/contract.rs:2]: Oops!

File: contracts/stream/src/helpers.rs [hunk 2]
```diff
@@ -55,3 +56,52 @@ pub fn check_name_and_url(name: &str, url: &Option<String>) -> Result<(), Contra
...
+        .seconds()
+        .checked_sub(now.seconds())
+        .ok_or(ContractError::StreamInvalidStartTime {})?;
+
+    let time_until_bootstrapping_start = bootstrapping_start_time
+        .seconds()
+        .checked_sub(now.seconds())
+        .ok_or(ContractError::StreamInvalidBootstrappingStartTime {})?;
+
+    if time_until_bootstrapping_start < params.min_seconds_until_bootstrapping_start_time {
```
Comment (orkunkl) [inline contracts/stream/src/helpers.rs:98]: Maybe operate on time duration between bootstrap and start_time?

File: contracts/stream/src/contract.rs [hunk 3]
```diff
@@ -125,50 +133,14 @@ pub fn execute(
...
-                Ok(execute_subscribe(
-                    deps,
-                    env,
-                    info,
-                    operator,
-                    operator_target,
-                    stream,
-                )?)
-            }
+            execute_subscribe(deps, env, info, operator, operator_target, stream)
```
Comment (orkunkl) [inline contracts/stream/src/contract.rs:136]: $$$

File: packages/types/src/stream/stream.rs [hunk 4]
```diff
@@ -1,69 +1,92 @@
 use crate::factory::CreatePool;
 use cosmwasm_schema::cw_serde;
-use cosmwasm_std::{Addr, Coin, Decimal, Decimal256, Timestamp, Uint128};
+use cosmwasm_std::{Addr, Coin, Decimal, Decimal256, Fraction, Timestamp, Uint128};
 use cw_vesting::msg::InstantiateMsg as VestingInstantiateMsg;
 use std::ops::Mul;
 
 #[cw_serde]
 pub struct Stream {
-    /// Name of the stream.
```
Comment (orkunkl) [inline packages/types/src/stream/stream.rs:9]: Uhh why delete these?

File: packages/types/src/stream/stream.rs [hunk 5]
```diff
@@ -1,69 +1,92 @@
...
     Waiting,
+    Bootstrapping,
     Active,
+    Ended,
     Finalized,
     Cancelled,
 }
 
+#[cw_serde]
+pub struct StreamStatus {
```
Comment (orkunkl) [inline packages/types/src/stream/stream.rs:38]: StatusInfo naming is more coherent

File: packages/types/src/stream/stream.rs [hunk 6]
```diff
@@ -1,69 +1,92 @@
...
+            return;
+        }
+        self.status = match now {
+            _ if now < self.bootstrapping_start_time => Status::Waiting,
+            _ if now >= self.bootstrapping_start_time && now < self.start_time => {
+                Status::Bootstrapping
+            }
+            _ if now >= self.start_time && now < self.end_time => Status::Active,
+            _ if now >= self.end_time => Status::Ended,
+            _ => self.status.clone(),
```
Comment (orkunkl) [inline packages/types/src/stream/stream.rs:73]:
```suggestion
            _ => unreachable!("This case should never occur"),
```

File: packages/types/src/stream/stream.rs [hunk 7]
```diff
@@ -1,69 +1,92 @@
...
+        StreamStatus {
+            status: Status::Waiting,
+            bootstrapping_start_time,
+            start_time,
+            end_time,
+            last_updated: now,
+        }
+    }
+
+    pub fn update_status(&mut self, now: Timestamp) {
```
Comment (orkunkl) [inline packages/types/src/stream/stream.rs:62]: Its best if we keep logic in contract folder

File: packages/types/src/stream/stream.rs [hunk 8]
```diff
@@ -73,42 +96,121 @@ impl Stream {
...
-            shares = (shares + self.in_supply - Uint128::one()) / self.in_supply;
+            (shares + self.in_supply - Uint128::one()) / self.in_supply
         } else {
-            shares /= self.in_supply;
+            shares / self.in_supply
         }
-        shares
+    }
+
+    pub fn update_status(&mut self, now: Timestamp) {
```
Comment (orkunkl) [inline packages/types/src/stream/stream.rs:125]: Best to keep this method either private or pub(crate)

File: tests/src/tests/streamswap_tests/subscribe.rs [hunk 9]
```diff
@@ -623,7 +633,166 @@ mod subscibe_test {
...
+            Decimal256::from_str("1406.292560801144492131").unwrap()
+        );
+        // Subscriber 2 has 150 in balance subscribed at 10th second
+        // Until 60 seconds, 60-10 = 50/90*150 = 83.333333333333333333
+        // 150-83.333333333333333333 = 66.666666666666666667
+        assert_eq!(position.in_balance, Uint128::new(66));
+        assert_eq!(position.spent, Uint128::new(84));
+    }
+    #[test]
+    fn test_subscibe_waiting() {
```
Comment (Ninjatosba) [inline tests/src/tests/streamswap_tests/subscribe.rs:675]: This test requires more attention as its a new status test

File: tests/src/tests/streamswap_tests/subscribe.rs [hunk 10]
```diff
@@ -511,7 +521,7 @@ mod subscibe_test {
         // Distribution index should not be updated because we are still in pending status
         assert_eq!(stream.dist_index, Decimal256::from_str("0").unwrap());
 
-        // Before stream start time, 2 subscriptions have been made and the stream is pending
+        // Before stream start time, 2 subscriptions have been made and the stream is bootstrapping
```
Comment (Ninjatosba) [inline tests/src/tests/streamswap_tests/subscribe.rs:524]: Follow this flow and calculations @orkunkl

File: packages/types/src/stream/stream.rs (hunk 4 above)
Comment (Ninjatosba) [reply packages/types/src/stream/stream.rs:9]: Ooops

File: contracts/stream/src/stream.rs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        return amount_in;
+    }
+    let shares = stream.shares.mul(amount_in);
+    if round_up {
+        (shares + stream.in_supply - Uint128::one()) / stream.in_supply
+    } else {
+        shares / stream.in_supply
+    }
+}
+pub fn update_stream(stream: &mut Stream, now: Timestamp) {
```
Comment (orkunkl) [inline contracts/stream/src/stream.rs:38]: lets make this sync_stream?

**Review by orkunkl — APPROVED** [review]
Great work!

**Pull request merged**